### PR TITLE
Support multiple object sections per XML file.

### PR DESCRIPTION
**Pull request by chkboom**
The current config only looks at the first <object></object> section and discards the rest.

This patch allows for more than one such sections per file. This comes in handy if using a plugin that groups things in profiles, for example. There's also the "input-device:..." settings which can be applied system-wide for instance on a custom piece of hardware. And best of all, no extra code is needed: just making the loop around create_section_from_xml_node.

I do notice the original code doesn't free the doc at the end (if successful), created here:
`auto doc = xmlParseFile(file.c_str());`
Is there a reason for this? Maybe worth mentioning it in a comment.

### CODE REVIEW

**Comment by ammen99** [comment]
I wanted to look at this PR but forgot, sorry. I do not understand what this brings to the table over just having multiple XML files? If you want to group settings together, this can be done with `<group>` tags which WCM supports. For the `.ini` file, it makes no difference.

From IRC:

```
I agree with you that the current config file format is far from optimal. Do you have any concrete proposals as to how to replace the xml + ini format, without losing current functionality?
the important things currently are:
1. Give plugins a way to describe their options (currently in XML) so that a config GUI like WCM is possible
2. Specify a type for each option
3. Specify a default value in one place (which is used by both WCM and wayfire)
```

**Comment by chkboom** [comment]
You can still have multiple XML files, however the current behaviour is to only read one of the object sections and stop without reading further. That actually seems like a bug.

For instance, a window decoration plugin which supports multiple themes.
```
<?xml version="1.0"?>
<wayfire>
	<plugin name="decor2">
		<_short>Decor II</_short>
		<_long>Advanced decoration for Wayfire windows.</_long>
		<category>Effects</category>
		<option name="border-size" type="int">
			<_short>Border size</_short>
			<_long>Border size in pixels</_long>
			<default>2</default>
		</option>
		<option name="border" type="color">
			<_short>Border color</_short>
			<_long>Color of the window border</_long>
			<default>#ffffffff</default>
		</option>
		<option name="title-back" type="color">
			<_short>Title background color</_short>
			<_long>Window title background color</_long>
			<default>#404040ff</default>
		</option>
		<option name="title-text" type="color">
			<_short>Title text color</_short>
			<_long>Window title text color</_long>
			<default>#ffffffff</default>
		</option>
	</plugin>
	<object name="decor2:active">
		<_short>Default decor2 theme (Active)</_short>
		<_long>Theme to apply when the window is active</_long>
		<option name="title-back" type="color"><default>#202040ff</default></option>
		<option name="title-text" type="color"><default>#ffffffff</default></option>
		<option name="border" type="color"><default>#000000ff</default></option>
	</object>
	<object name="decor2:inactive">
		<_short>Default decor2 theme (Inactive)</_short>
		<_long>Theme to apply when the window is inactive</_long>
		<option name="title-back" type="color"><default>#202020c0</default></option>
		<option name="title-text" type="color"><default>#ffffffff</default></option>
		<option name="border" type="color"><default>#00000080</default></option>
	</object>
</wayfire>
```
The corresponding text in the wayfire.ini would look like this:
```
[decor2]
border-size = 2

[decor2:active]
title-back = #404040c0
title-text = #ffffffee
border = #00000080

[decor2:inactive]
title-back = #202020a0
title-text = #eeeeeeee
border = #00000070
```
If I had this in a single decor2.xml file, wf-config would not read beyond the first "plugin" tag. The only work around is to have three XML files just for one plugin.
This patch fixes that and parses all the plugin/object tags that may be in the one file, instead of just stopping after reading the first one.

**Comment by ammen99** [comment]
In your example, I think you misunderstand what objects are. They are templates for sections, not predefined sections. They were added for things like output configuration: you have [output:<xxx>] object for each output. The <xxx> is not predefined. If you want a section in the config file with a fixed name, you should use <plugin>.

I nevertheless see your point about multiple <plugin>s in a single .xml file. I have two requests before merging:

- Please rebase on top of master
- Retain the previous code structure, e.g. use a helper function for processing a single .XML file. This is simply to avoid having too much nested logic.

**Comment by chkboom** [comment]
> In your example, I think you misunderstand what objects are. They are templates for sections, not predefined sections. They were added for things like output configuration: you have [output:] object for each output. The is not predefined. If you want a section in the config file with a fixed name, you should use .

I think parts of your message went missing. "The is not predefined." "you should use ."
Do object and section have different behaviors or is that mainly a semantic thing?

> I nevertheless see your point about multiple s in a single .xml file. I have two requests before merging:
> 
>     * Please rebase on top of master
I tried that, though I'm not sure I was successful. It pulled in the latest commits from master, is that what you were after? Never done rebasing in git before though. Ended up re-triggering Github the workflow check, waiting for approval now.

>     * Retain the previous code structure, e.g. use a helper function for processing a single .XML file. This is simply to avoid having too much nested logic.
Done; also added a comment as to why xmlFreeDoc(doc) is not called at the end.

**Review by ammen99 — APPROVED** [review]
LGTM now, thanks!

I really don't know why I did not get a notification when you updated this PR last year ..
Anyway, yes, objects and normal plugin sections are meant to behave differently. Objects are more like object templates, think of Wayfire's outputs: We have a single object (template) called `output` and it is 'instantiated' for each output: so you have sections `[output:HDMI-A-1]` etc. A `<plugin>` on the other hand is static, there are no 'instances': there is only one `[move]` plugin. I suggested using normal `<plugin>`s in your case, because you have a fixed amount of options, but on the other hand, you'd have to duplicate the XML sections .. A third option would be to use compound options, but they are even more cumbersome in the `.ini` file. EDIT: I see with objects, you already have duplicate code in the xml file, so it would indeed be better to use `<plugin>` since you don't do any instantiation.

Of course, in the end, you can use either of them for your plugin. I think the confusion is partly due to choice of the ini file format, which in retrospect is very inadequate for our use-case.

**Pull request merged**
